scheduler CC: add v1beta2 API, deprecate plugins

adtac opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind api-change

#### What this PR does / why we need it:

Adds scheduler config v1beta2 API. This PR builds on top of https://github.com/kubernetes/kubernetes/pull/95308.

xref: https://github.com/kubernetes/kubernetes/issues/95446 - things completed in this PR:

* [x]  Deprecate some Plugins #94008
* [ ]  Deprecate resources plugins in favor of a unified Fit plugin #98746 (not doing this in this PR)
* [x]  Remove backwards compatibility hacks:
  * [x]  Always set DefaultingType to System (PodTopologySpread). Hack added in #95048
  * [x]  Stop swallowing invalid healthz and metrics address #97048
* [ ]  [Nice-to-have] Make Plugin names fully qualified #95445 (not doing this in this PR)

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related to https://github.com/kubernetes/kubernetes/issues/94008

#### Special notes for your reviewer:

Commits in @hprateek43's PR are preserved as-is. See everything after the 3rd commit for the changes I've made on top.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-scheduler component config v1beta2 API available
Three scheduler plugins deprecated (NodeLabel, ServiceAffinity, NodePreferAvoidPods)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

cc @hprateek43 @alculquicondor @ahg-g @liggitt

adtac commented on cmd/kube-scheduler/app/options/options_test.go:
```diff
@@ -373,7 +443,8 @@ profiles:
 					RemoteKubeConfigFileOptional: true,
-					AlwaysAllowPaths:             []string{"/healthz"}, // note: this does not match /healthz/ or /healthz/*
+					AlwaysAllowPaths:             []string{"/healthz", "/readyz", "/livez"}, // note: this does not match /healthz/ or /healthz/*
+					AlwaysAllowGroups:            []string{"system:masters"},
```
@hprateek43 I changed this back to make sure v1beta1 and v1beta2 tests were similar. Was there a reason for your change that I missed?

adtac commented on pkg/scheduler/apis/config/v1beta2/defaults.go:
```diff
@@ -55,14 +53,14 @@ func SetDefaults_KubeSchedulerConfiguration(obj *v1beta2.KubeSchedulerConfigurat
 	// Only apply a default scheduler name when there is a single profile.
 	// Validation will ensure that every profile has a non-empty unique name.
 	if len(obj.Profiles) == 1 && obj.Profiles[0].SchedulerName == nil {
-		obj.Profiles[0].SchedulerName = pointer.StringPtr(api.DefaultSchedulerName)
```
@hprateek43 did you mean to leave it as v1.DefaultSchedulerName? (k8s.io/kubernetes/pkg/apis/core.DefaultSchedulerName) is undefined.

alculquicondor replied:
that was removed recently

adtac reviewed (commented):
Just to double-check, I did a diff of all files in v1beta1 and v1beta2 and here's a list of all differences minus v1beta1/v1beta2 differences: http://ix.io/2RjG/diff

adtac commented:
/sig scheduling

alculquicondor commented on cmd/kube-scheduler/app/options/options_test.go:
```diff
@@ -34,11 +34,26 @@ import (
 	componentbaseconfig "k8s.io/component-base/config"
+	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
 	"k8s.io/component-base/logs"
+	configv1beta1 "k8s.io/kube-scheduler/config/v1beta1"
```
nit: can we have these without aliases?
Or at least the aliases don't have to say "kubescheduler", given the location of the file, if there are no conflicts.

alculquicondor commented on cmd/kube-scheduler/app/options/options_test.go:
```diff
@@ -272,7 +342,7 @@ profiles:
 	}{
 		{
-			name: "config file",
+			name: "config file", // v1beta2
```
any reason not to make v1beta2 part of the test name?

alculquicondor replied:
better to have more context when I'm running the test.

alculquicondor commented on pkg/scheduler/apis/config/v1beta2/conversion.go:
```diff
@@ -46,6 +46,98 @@ func getPluginArgConversionScheme() *runtime.Scheme {
 	return pluginArgConversionScheme
 }
 
+func Convert_v1beta2_Plugins_To_config_Plugins(in *v1beta2.Plugins, out *config.Plugins, s conversion.Scope) error {
```
What if we get rid of pointers?

adtac replied:
do you mean `type Plugins struct { QueueSort PluginSet }`? instead of `*PluginSet` in the v1beta2 API?

alculquicondor replied:
correct

adtac replied:
Done. (See fourth commit.)

alculquicondor commented on pkg/scheduler/apis/config/v1beta2/defaults.go:
```diff
@@ -80,8 +78,9 @@ func SetDefaults_KubeSchedulerConfiguration(obj *v1beta2.KubeSchedulerConfigurat
 			} else {
-				// TODO: in v1beta2 we should let this error instead of stomping with a default value
-				obj.HealthzBindAddress = &defaultBindAddress
+				// This is a malformed partial address and there's nothing we can do
```
Isn't there any guards in validation.go?
Also I don't think the comment is necessary.

adtac replied:
You're right, it is caught by validation. However we didn't test this scenario in validation_test.go, so I've added a test for this.

alculquicondor replied:
awesome

alculquicondor commented on pkg/scheduler/apis/config/v1beta2/defaults.go:
```diff
@@ -211,7 +211,11 @@ func SetDefaults_VolumeBindingArgs(obj *v1beta2.VolumeBindingArgs) {
 	if feature.DefaultFeatureGate.Enabled(features.DefaultPodTopologySpread) {
 		if obj.DefaultingType == "" {
-			obj.DefaultingType = v1beta2.SystemDefaulting
+			if len(obj.DefaultConstraints) != 0 {
```
why revert this? It was intended. Defaults should be independent of other fields.

adtac replied:
Oops, good catch, I missed this when reviewing.

alculquicondor reviewed (commented):
Looking good. Could you start a PR for #94008 already?

alculquicondor commented on cmd/kube-scheduler/app/options/options_test.go:
```diff
@@ -34,21 +34,21 @@ import (
+	"k8s.io/kube-scheduler/config/v1beta1"
 	kubeschedulerconfig "k8s.io/kubernetes/pkg/scheduler/apis/config"
-	kubeschedulerscheme "k8s.io/kubernetes/pkg/scheduler/apis/config/scheme"
+	scheme "k8s.io/kubernetes/pkg/scheduler/apis/config/scheme"
```
remove the alias when it matches the package name.

Commit:
scheduler: graduate CC to v1beta2, deprecate plugins

Signed-off-by: Adhityaa Chandrasekar <adtac@google.com>

adtac commented:
I've squashed some commits and added a new commit. The second commit in this PR adds the infrastructure to deprecate plugins in CC; in v1beta2, we want to deprecate NodeLabel and ServiceAffinity. The second commit was a bit tricky from a API point-of-view, so please lmk if I'm doing something wrong.

liggitt commented:
> I know we're past the 1.21 code freeze, so we'll be targetting v1beta2 in 1.22 (with possibly more changes than there are in this PR)

sounds good

/milestone v1.22

alculquicondor commented on pkg/scheduler/apis/config/types_pluginargs.go:
```diff
@@ -56,7 +56,9 @@ type InterPodAffinityArgs struct {
-// NodeLabelArgs holds arguments used to configure the NodeLabel plugin.
+// NodeLabelArgs holds arguments used to configure the NodeLabel plugin. Note:
+// this plugin has been deprecated and is only configurable through the
+// scheduler policy API.
```
but also v1beta1, isn't it?

alculquicondor commented on pkg/scheduler/apis/config/validation/validation.go:
```diff
@@ -80,6 +83,80 @@ func ValidateKubeSchedulerConfiguration(cc *config.KubeSchedulerConfiguration) f
+
+func validateProfilesForRemovedPlugins(path *field.Path, apiVersion string, profiles []config.KubeSchedulerProfile) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for i, profile := range profiles {
```
shouldn't this be part of `validateKubeSchedulerProfile` instead?

adtac replied:
Good point, moved it inside that function

alculquicondor commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -402,13 +412,21 @@ func registerClusterEvents(name string, eventToPlugins map[framework.ClusterEven
+	switch componentConfigVersion {
+	case v1beta1.SchemeGroupVersion.String():
+		gvk = v1beta1.SchemeGroupVersion.WithKind(name + "Args")
+	case v1beta2.SchemeGroupVersion.String():
```
we should do empty equivalent to v1beta2, otherwise we have to specify the version in every test.

adtac commented:
@liggitt this is ready for API review

ahg-g commented:
/label api-review

alculquicondor commented on pkg/scheduler/apis/config/types_pluginargs.go:
```diff
@@ -180,7 +182,9 @@ type ResourceSpec struct {
 
-// ServiceAffinityArgs holds arguments used to configure the ServiceAffinity plugin.
+// ServiceAffinityArgs holds arguments used to configure the ServiceAffinity
+// plugin. Note: this plugin has been deprecated and is only configurable
```
nit: https://blog.golang.org/godoc#:~:text=Deprecated

adtac replied:
so I reverted this change because of staticcheck, which checks for uses of symbols that have "Deprecated:" in their docs. Since we still use these types in a lot of places in the scheduler code, I think it makes sense to back to not using "Deprecated:". WDYT?

alculquicondor reviewed (commented):
It looks like you missed one plugin from the release note.

Can you list in the PR description what things are being done here from #95446?

adtac commented on cmd/kube-scheduler/app/options/configfile.go:
```diff
@@ -37,32 +37,32 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
 	}
 
-	encoder := kubeschedulerscheme.Codecs.EncoderForVersion(info.Serializer, kubeschedulerconfigv1beta1.SchemeGroupVersion)
+	encoder := scheme.Codecs.EncoderForVersion(info.Serializer, configv1beta2.SchemeGroupVersion)
```
Done. I've also modified the tests to test this scenario.

alculquicondor reviewed (commented):
Include all the relevant changes in the release note.

adtac commented on cmd/kube-scheduler/app/options/configfile.go:
```diff
@@ -37,39 +40,55 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
+	default:
+		return buf, fmt.Errorf("unexpected component config version %q", cfg.ComponentConfigVersion)
+	}
+	if err := encoder.Encode(cfg, buf); err != nil {
```
runtime.Encode (returns []byte) uses encoder.Encode (returns bytes.Buffer) internally; it's just a small wrapper. I just liked the bytes.Buffer interface better than dealing with raw []byte

adtac commented on pkg/scheduler/apis/config/v1beta2/defaults.go:
```diff
@@ -0,0 +1,213 @@
+			}
+			hostPort := net.JoinHostPort(host, port)
+			obj.HealthzBindAddress = &hostPort
+		} else {
```
No, what we wanted to stop doing was this:

https://github.com/kubernetes/kubernetes/blob/5b46064cae41395cf775cb445ffdaae1261076c7/pkg/scheduler/apis/config/v1beta1/defaults.go#L79-L82

The else clause in this PR is there so that plain IPs are allowed in addition to IP:port. We no longer allow inputs that are invalid IP and invalid IP:port, which is what that issue is talking about.

I'll update the release note and mark the completed items in a bit.

alculquicondor replied:
ahh, it was the next else. Sorry :)

adtac commented:
@alculquicondor I've updated the PR description with a checklist of things done in this PR. Also, I just realised I had already deprecated NodePreferAvoidPods in this PR but I forgot about it :sweat_smile: (context: we discussed this offline recently)

alculquicondor commented on pkg/scheduler/framework/plugins/nodelabel/node_label.go:
```diff
@@ -14,6 +14,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// Deprecated: this plugin has been deprecated and is only configurable through
```
maybe we can also add a log line warning about the use of a deprecated plugin in v1beta1.

adtac replied:
ack. I've kept the message simple.

alculquicondor commented:
/approve

please squash

adtac commented on pkg/scheduler/apis/config/v1beta2/defaults.go:
```diff
@@ -0,0 +1,213 @@
+	// 1. If the value is nil, default to 0.0.0.0 and default scheduler port
+	// 2. If there is a value set, attempt to split it. If it's just a port (ie, ":1234"), default to 0.0.0.0 with that port
+	// 3. If splitting the value fails, check if the value is even a valid IP. If so, use that with the default port.
+	// Otherwise leave the address as-is, it will be caught during validation.
```
good catch, added a test (see "metrics and healthz with invalid IP" in defaults_test.go)

adtac commented on pkg/scheduler/apis/config/validation/validation.go:
```diff
@@ -80,16 +82,57 @@ func ValidateKubeSchedulerConfiguration(cc *config.KubeSchedulerConfiguration) u
+	},
+}
+
+func isPluginRemoved(componentConfigVersion string, name string) (bool, string) {
```
>behavior for future versions

could you elaborate?

msau42 replied:
I think the comment on `removedPlugins` addresses this thanks!

liggitt commented on pkg/scheduler/apis/config/types.go:
```diff
@@ -49,6 +49,10 @@ const (
 type KubeSchedulerConfiguration struct {
 	metav1.TypeMeta
 
+	// ComponentConfigVersion is the external type version that should be used to
```
*if* you're going to set a field in conversion (which is unusual), is there a reason not to set the TypeMeta.APIVersion field which already exists to hold this info?

adtac replied:
So I did consider TypeMeta, but I was honestly expecting it to be filled by API machinery - is that not the case? (TypeMeta.APIVersion was empty)

If TypeMeta.APIVersion is not apimachinery-reserved or anything like that, I agree with you

liggitt replied:
When converting from external → internal, the internal object is not expected to bear any trace of the originating external version, so APIVersion is cleared.

This is explicitly going against that convention by adding a new internal field specifically to track the originating external version. If it's going to do that, I can't think of a reason not to use the field already present to hold APIVersion info

ahg-g commented:
Are we close to merge this PR? do you need help moving it forward?

msau42 commented on pkg/scheduler/apis/config/types_pluginargs.go:
```diff
@@ -57,6 +57,10 @@ type InterPodAffinityArgs struct {
+//
+// This plugin has been deprecated and is only configurable through the
+// scheduler policy API and the v1beta1 component config API. It is recommended
+// to use the NodeAffinity plugin instead.
```
Can we also add this deprecation notice on the v1beta1 api?

msau42 reviewed (commented):
From an API standpoint, the main thing that is remaining is to resolve https://github.com/kubernetes/kubernetes/pull/99597/files#r631975169

ahg-g commented:
> From an API standpoint, the main thing that is remaining is to resolve https://github.com/kubernetes/kubernetes/pull/99597/files#r631975169

It seems Adi is having some issues resolving this. Can we address it in a followup PR? This is not a user facing change, so it is safe to address later. I would like to get this PR merged asap because it is blocking multiple followup work.

ahg-g commented:
> > From an API standpoint, the main thing that is remaining is to resolve https://github.com/kubernetes/kubernetes/pull/99597/files#r631975169
> 
> It seems Adi is having some issues resolving this. Can we address it in a followup PR? This is not a user facing change, so it is safe to address later. I would like to get this PR merged asap because it is blocking multiple followup work.

@liggitt are you ok with this?

adtac commented:
I think I've solved all issues now. PTAL.

tl;dr of what was blocking: API machinery seems to overwrite the TypeMeta.APIVersion field every time a conversion happens here:

https://github.com/kubernetes/kubernetes/blob/12e3860cb084ef99f79eb2bc7c123f9521342aff/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L544-L546

This makes sense in general, but we want something slightly different, so instead setting the APIVersion field in conversion.go (which doesn't work), we set it after we read the config file in post-conversion.

ahg-g commented:
@liggitt I think Adi now addressed all the comments, please take a look.

msau42 reviewed (commented):
latest api changes lgtm

liggitt commented on cmd/kube-scheduler/app/options/configfile.go:
```diff
@@ -37,39 +40,61 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
+	}
+
+	var encoder runtime.Encoder
+	switch cfg.TypeMeta.APIVersion {
```
this is switching on encoding the internal config based on the external version it has in its apiVersion field? shouldn't we pick a default version to encode to (latest seems reasonable) like we did previously?

adtac replied:
since the internal object could've been converted from any version, there could be some information loss if we convert for example v1beta1 -> internal -> v1beta2.

to be quite honest, and I discussed this with @alculquicondor, I don't understand the purpose of this feature. who is using this and for what?

liggitt replied:
> I don't understand the purpose of this feature. who is using this and for what?

isn't this the path taken when generating a config with `--write-config-to`?

adtac replied:
Yes.

adtac replied:
right, but I don't fully understand what that feature is used for. perhaps to generate a default config file that you can modify afterwards?

liggitt replied:
> perhaps to generate a default config file that you can modify afterwards?

it's not just a default config file, it writes the effective configuration resulting from the flags you set, so someone transitioning from flags to a config file can add `--write-config-to=foo.yaml` to their existing flag-based invocation and get the equivalent config file.

alculquicondor replied:
using latest version sgtm

adtac replied:
Done.

adtac replied:
actually, I'm having second thoughts about this: we necessarily need to switch when someone loads a v1beta1 config file that uses a deprecated plugin because the v1beta2 encoder cannot encode plugin args for deprecated plugins. am I missing something?

liggitt replied:
I don't object to switching on APIVersion, my comment was limited to the behavior when APIVersion was empty (pick a default, instead of erroring)

adtac replied:
ah I completely misread your comment. done.

liggitt commented on cmd/kube-scheduler/app/options/configfile.go:
```diff
@@ -37,39 +40,61 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
+		// conversion. To work around that, we set the API version field of the
+		// internal object here to the version we read from file. We need this for
+		// later when we need to make decisions based on the external version.
+		cfgObj.TypeMeta.APIVersion = gvk.GroupVersion().String()
```
do you have a pointer to the decisions made based on this field? it's pretty strange for the external version source to change the interpretation of the converted internal config

liggitt replied:
if the only decision made is a pass/fail validation decision, that would be helpful to clarify here. we explicitly don't want the internal configuration getting interpreted differently based on version

ahg-g replied:
There are two places, the validation and the plugin config defaulting logic inside the framework. The former will continue to exist, the latter will be removed as a followup to this PR (https://github.com/kubernetes/kubernetes/issues/102211).

adtac replied:
*three places. We also use this in `--write-config-to`. I've added a comment in the internal KubeSchedulerConfiguration with references to exactly what places use this field.

liggitt commented on pkg/scheduler/framework/runtime/framework.go:
```diff
@@ -415,13 +425,22 @@ func registerClusterEvents(name string, eventToPlugins map[framework.ClusterEven
+		gvk = v1beta1.SchemeGroupVersion.WithKind(name + "Args")
+	case v1beta2.SchemeGroupVersion.String():
+		gvk = v1beta2.SchemeGroupVersion.WithKind(name + "Args")
+	default:
```
should we default to this even if component config version is non-empty and is an unknown version?

ahg-g replied:
The original code was doing it regardless of `componentConfigVersion`, I think this is fine as is. In any case, this defaulting logic will be removed as a followup as part of https://github.com/kubernetes/kubernetes/issues/102211

liggitt commented on pkg/scheduler/scheduler.go:
```diff
@@ -110,6 +111,14 @@ type schedulerOptions struct {
 type Option func(*schedulerOptions)
 
+// WithComponentConfigVersion sets the component config version for the
+// KubeSchedulerConfiguration used.
```
document examples to make it clear this is the full group/version

liggitt commented on pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go:
```diff
@@ -70,6 +74,7 @@ func New(plArgs runtime.Object, handle framework.Handle) (framework.Plugin, erro
 	}
 	serviceLister := handle.SharedInformerFactory().Core().V1().Services().Lister()
 
+	klog.Warning("ServiceAffinity plugin is deprecated and will be removed in future versions")
```
put the recommended replacement (InterPodAffinity) in the user-facing warning

liggitt commented on pkg/scheduler/framework/plugins/nodepreferavoidpods/node_prefer_avoid_pods.go:
```diff
@@ -85,5 +89,6 @@ func (pl *NodePreferAvoidPods) ScoreExtensions() framework.ScoreExtensions {
 
 // New initializes a new plugin and returns it.
 func New(_ runtime.Object, h framework.Handle) (framework.Plugin, error) {
+	klog.Warning("NodePreferAvoidPods plugin is deprecated and will be removed in future versions")
```
put the recommended replacement (node taints) in the user-facing warning

liggitt commented on pkg/scheduler/framework/plugins/nodelabel/node_label.go:
```diff
@@ -46,6 +50,7 @@ func New(plArgs runtime.Object, handle framework.Handle) (framework.Plugin, erro
 		return nil, err
 	}
 
+	klog.Warning("NodeLabel plugin is deprecated and will be removed in future versions")
```
mention the NodeAffinity plugin as a replacement

adtac commented on cmd/kube-scheduler/app/options/options.go:
```diff
@@ -133,14 +133,20 @@ func splitHostIntPort(s string) (string, int, error) {
+	// because the field will be cleared later by API machinery during
+	// conversion. To work around that, we set the API version field of the
+	// internal object here to the default version (latest). We need this for
+	// later when we need to make decisions based on the external version.
```
That is correct. I've added a comment in the internal KubeSchedulerConfiguration type definition.

liggitt commented:
/approve

ahg-g commented:
/hold

pls squash :)

ahg-g commented:
/lgtm

ahg-g commented:
/hold cancel

adtac commented:
/retest
